### PR TITLE
Add stump-nm, a StumpWM and NetworkManager integration module.

### PR DESCRIPTION
**Pull request by ralt**
This module allows you to manage your Wi-Fi networks and VPN
connections from within StumpWM itself. It is intentionally pretty
bare-bones in features, in that it allows you to enable/disable
connections, and no more. It is not a replacement for `nmtui` and/or
`nmcli`.

It offers 2 commands:

- `nm-list-wireless-networks`
- `nm-list-vpn-connections`

Each command offers a menu where you can select a WiFi or VPN
connection.

More information in the README.org.

### CODE REVIEW

File: util/stump-nm/nm-api.lisp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                                                               active-access-point)))
+                  (setf (ssid ap) (access-point-ssid ap)
+                        (frequency ap) (getprop ap "Frequency")
+                        (strength ap) (getprop ap "Strength"))))
+              (object (interface device) "GetAllAccessPoints")))))
+
+(defun access-point-ssid (ap)
+  (format nil "~{~a~}"
+          (mapcar #'code-char
+                  (getprop ap "Ssid"))))
```
Comment (PuercoPop) [inline util/stump-nm/nm-api.lisp:95]: iiuc this code first transforms octets to chars and then collects them in a string, if so you can do the same by using `map` and specifying string as the result-type.
```suggestion
(defun access-point-ssid (ap)
  (map 'string #'code-char (getprop ap "Ssid")))
```

However this approach only works for code-points that fit in one 8-bit byte. ej.
```lisp
(let ((octets #(227 130 179 227 131 162 227 131 179 227 131 170 227 130 185 227 131 151 227 129 140 227 129 153 227 129 141 227 129 167 227 129 153)))
  (map 'string #'code-char octets)) ; => "ã³ã¢ã³ãªã¹ããããã§ã"
```

To remedy this please use the helper `stumpwm::utf8-to-string` which is a thin wrapper over `sb-ext:octets-to-string`.


```lisp
(let ((octets #(227 130 179 227 131 162 227 131 179 227 131 170 227 130 185 227 131 151 227 129 140 227 129 153 227 129 141 227 129 167 227 129 153)))
  (stumpwm::utf8-to-string octets)) ; => "コモンリスプがすきです"
```


sb-ext:octets-to-string or the wrapper we provide `utf8-to-string`

File: util/stump-nm/package.lisp [hunk 2]
```diff
@@ -0,0 +1,3 @@
+(defpackage :stump-nm
+  (:use :cl :alexandria :dbus)
+  (:export :nm-list-wireless-networks :nm-list-vpn-connections))
```
Comment (PuercoPop) [inline util/stump-nm/package.lisp:3]:
```suggestion
(defpackage #:stump-nm
  (:use #:cl #:alexandria #:dbus)
  (:export #:nm-list-wireless-networks #:nm-list-vpn-connections))
```

File: util/stump-nm/nm-api.lisp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+   (uuid :accessor uuid)
+   (state :accessor state)))
+
+(defun get-active-connections (root)
+  (mapcar (lambda (connection-path)
+            (make-active-connection root connection-path))
+          (getprop root "ActiveConnections")))
+
+(defun make-active-connection (root connection-path)
+  (prog1-let (active-connection (make-instance
```
Comment (PuercoPop) [inline util/stump-nm/nm-api.lisp:108]: If you don't the extra dependency you can use serapeum's lret for this https://github.com/ruricolist/serapeum/blob/master/REFERENCE.md#lret-rest-bindings-body-body

File: util/stump-nm/stump-nm.asd [hunk 4]
```diff
@@ -0,0 +1,9 @@
+(defsystem "stump-nm"
+  :serial t
+  :description "StumpWM integration with NetworkManager"
+  :author "Florian Margaine <florian@margaine.com>"
+  :license "GPLv3"
+  :depends-on (:alexandria :stumpwm :dbus)
```
Comment (PuercoPop) [inline util/stump-nm/stump-nm.asd:6]:
```suggestion
  :depends-on ("alexandria" "stumpwm" "dbus")
```

**Review by PuercoPop — COMMENTED** [review]
Hey thanks for this! :tada: 

Left a couple of suggestions. The important one is to use utf8-to-string. I'd rather you drop the get- prefix as it doesn't adds any information imho but it is not a hill I'm willing to die on.

Another nit for the future is to use upper case for the format directives (ej. `~A`) but don't worry about it for this PR.

One last thing, can you run the `./update-readme.sh` script and commit the updated readme as well?

File: util/stump-nm/nm-api.lisp (hunk 1 above)
Comment (ralt) [reply util/stump-nm/nm-api.lisp]: Very good point! I've used babel instead, it's already loaded and OCTETS-TO-STRING is actually exported.

File: util/stump-nm/nm-api.lisp (hunk 3 above)
Comment (ralt) [reply util/stump-nm/nm-api.lisp:110]: I mind a bit, especially for such a big dependency.

**Comment by ralt** [comment]
Hey @PuercoPop thanks a lot for the review! I've pushed a new version following your feedback. I also edited the docstring of GETPROP to be a bit clearer on the naming convention.

**Pull request merged**